### PR TITLE
api generator

### PR DESCRIPTION
**Pull request by drujensen**
### Description of the Change

Provide the ability to generate an API endpoint with model and migration.  

### Alternate Designs

none

### Benefits

Easily scaffold an API for quick POC's

### Possible Drawbacks

This is a working version but there are several improvements needed IMO:
1. We need to be able to support `amber g api Api::Post title:string` to provide namespaces.
2. We need to provide a simpler `responds_with` option for `json`
3. The specs need some improvement to check the DB for actual changes.

#### Setup
```
amber new blog --deps
cd blog
amber g api post title:string
amber db drop create migrate
amber w
```

#### Index
```
curl \
  -i 'http://localhost:3000/posts' \
  -X GET \
  -H "Content-Type:application/json" \
  -H 'Accept: application/json' \
  -H "Origin:localhost"
```

#### Show
```
curl \
  -i 'http://localhost:3000/posts/1' \
  -X GET \
  -H "Content-Type:application/json" \
  -H 'Accept: application/json' \
  -H "Origin:localhost"
```

#### Create
```
curl \
  -i 'http://localhost:3000/posts' \
  -X POST \
  -H "Content-Type:application/json" \
  -H 'Accept: application/json' \
  -H "Origin:localhost" \
  -d '{"title": "New Blog Post"}'
```

#### Update
```
curl \
  -i 'http://localhost:3000/posts/1' \
  -X PATCH \
  -H "Content-Type:application/json" \
  -H 'Accept: application/json' \
  -H "Origin:localhost" \
  -d '{"title": "Old Blog Post"}'
```

#### Destroy
```
curl \
  -i 'http://localhost:3000/posts/1' \
  -X DELETE \
  -H "Content-Type:application/json" \
  -H 'Accept: application/json' \
  -H "Origin:localhost"
```

### CODE REVIEW

**Comment by damianham** [comment]
I would have thought recipes were the best way to provide an API application template.

**Comment by faustinoaq** [comment]
> I would have thought recipes were the best way to provide an API application template.

Yeah, I have been contributing and testing recipes. An API recipe would full fill this use-case in a very nice way, although, granite and crecto templates are already here, so maybe we can have a `basic/api` recipe or even just `api` recipe and keep this. @damianham WDYT?

PS: Let me try to create an API recipe based on this PR :wink: Just like I did with Jennifer :+1: 

BTW, recipes are easier to test (all use-cases combinations are being tested, see: [recipes/specs](https://github.com/amberframework/recipes/tree/master/spec) I found and fixed a lot of bugs with this), and auto-deployed, so maybe in the future we can remove all amber embedded templates and generators and just use recipes :sweat_smile: 

Also, I just realize amber templates already have a lot of duplicated code, recipes can solve that by using  hierarchical recipes, see: https://github.com/amberframework/recipes/issues/16

**Comment by drujensen** [comment]
@damianham @faustinoaq I think you guys are thinking of a different use case than I am.

I would like to generate an API and a WEB interface in the same application.  Amber is well suited for this using pipelines.   I am not looking for a new application that is API only.  Does that make sense?

**Comment by drujensen** [comment]
@faustinoaq If you want to work on an API only application, then I think building a recipe is fine but that will not help my use case.  I would still want this API generator.

**Comment by damianham** [comment]
@drujensen a recipe is still the best way to do that.  The react/preact_redux recipe does exactly that (if you remove the React JS stuff) although it doesn't use different pipelines it uses respond_with to respond with either html or json.  If you are creating a new api generator in the CLI because you want to be able to add routes to 2 pipelines during scaffolding - modules and routes defiend in modules are the way to solve that problem :-)
@faustinoaq I think an api recipe at the root rather than basic/api.

**Comment by drujensen** [comment]
>> by creating an internal api generator you are just duplicating the scaffold generator but with JSON controller responses and without the views.

It's a little more than that since we will be introducing an `api` pipeline as well.  Also the routes `resources` will need to be changed to remove the `new` and `edit` routes.

Eventually, we will be moving the recipes into their own repos, so your example:
`amber new my_api_app -r api` will be `amber new my_api_app -r damianahm/api` but close enough.

**Comment by drujensen** [comment]
>> Also I see this is limited to granite or crecto, thus limiting support for other ORMs

This is true and something the core team has been discussing.  We might remove the `-m` flag and use recipes to handle different ORM's.  It's still something we have been debating.

**Comment by faustinoaq** [comment]
> sorry @faustinoaq after all your excellent work with recipes

No problem, Thank you for your new proposal :smile: :+1:

File: src/amber/cli/templates/api/controller/granite/src/controllers/{{name}}_controller.cr.ecr [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    end
+  end
+
+  def create
+    <%= @name %> = <%= class_name %>.new(<%= @name %>_params.validate!)
+
+    if <%= @name %>.valid? && <%= @name %>.save
+      responds_with 201 do
+        json <%= @name %>.to_json
+      end
```
Comment (robacarp) [inline src/amber/cli/templates/api/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: These `responds_with` blocks seem verbose given that it's always going to respond with json, is that the only way to render json with Amber?

Comment (robacarp) [reply src/amber/cli/templates/api/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: Actually, shouldn't these be `respond_with` (singular)?

Comment (drujensen) [reply src/amber/cli/templates/api/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: still a WIP but yeah.

File: src/amber/cli/templates/api/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    response = subject.get "/<%= Inflector.pluralize(@name) %>"
+
+    response.status_code.should eq(200)
+    response.body.should contain("Fake")
+  end
+
+  it "renders <%= @name %> show json" do
+    <%= class_name %>.clear
+    model = create_<%= @name.downcase %>
+    location = "/<%= Inflector.pluralize(@name) %>/#{model.id}"
```
Comment (robacarp) [inline src/amber/cli/templates/api/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr]: I see this snippet everywhere. Does it make sense to put the plural in the teeplate controller above?

Comment (drujensen) [reply src/amber/cli/templates/api/controller/granite/spec/controllers/{{name}}_controller_spec.cr.ecr]: @robacarp Yes, yes it does.  👍

**Comment by robacarp** [comment]
@elorest that seems like it'd be a handy utility, but I think it's outside the scope of this generator pull

File: src/amber/cli/templates/api/controller.cr [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      when "float32", "float64", "float"
+        "1.00"
+      when "bool", "boolean"
+        "true"
+      when "time", "timestamp"
+        Time.now.to_s
+      when "ref", "reference", "references"
+        rand(100).to_s
+      else
+        "Fake"
```
Comment (eliasjpr) [inline src/amber/cli/templates/api/controller.cr]: Why not just an empty string here?

File: src/amber/cli/templates/api/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  end
+end
+
+describe <%= class_name %>ControllerTest do
+  subject = <%= class_name %>ControllerTest.new
+
+  it "renders <%= @name %> index json" do
+    <%= class_name %>.clear
+    model = create_<%= @name.downcase %>
+    response = subject.get "/<%= Inflector.pluralize(@name) %>"
```
Comment (eliasjpr) [inline src/amber/cli/templates/api/controller/crecto/spec/controllers/{{name}}_controller_spec.cr.ecr]: The inflector is used a lot here. Maybe we can find a way to put this `Inflector.pluralize(@name)` in a higher order and just pass a variable to the template? 

```crystal
resource_path = "/#{Inflector.pluralize(@name)}"
```

**Comment by elorest** [comment]
> @elorest that seems like it'd be a handy utility, but I think it's outside the scope of this generator pull

@robacarp I'm fine with merging without that option. Just pointing it out as rails scaffolding allows you to skip models and also asks if we should overwrite existing files.

**Pull request merged**